compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java

```diff
@@ -899,24 +899,79 @@ public void visit(BLangArrayType arrayTypeNode) {
         if (resultType == symTable.noType) {
             return;
         }
+        boolean isError = false;
         for (int i = 0; i < arrayTypeNode.dimensions; i++) {
             BTypeSymbol arrayTypeSymbol = Symbols.createTypeSymbol(SymTag.ARRAY_TYPE, Flags.PUBLIC, Names.EMPTY,
                                                                    env.enclPkg.symbol.pkgID, null, env.scope.owner);
             BArrayType arrType;
             if (arrayTypeNode.sizes.length == 0) {
                 arrType = new BArrayType(resultType, arrayTypeSymbol);
             } else {
-                int size = arrayTypeNode.sizes[i];
-                arrType = (size == UNSEALED_ARRAY_INDICATOR) ?
-                        new BArrayType(resultType, arrayTypeSymbol, size, BArrayState.UNSEALED) :
-                        (size == OPEN_SEALED_ARRAY_INDICATOR) ?
-                                new BArrayType(resultType, arrayTypeSymbol, size, BArrayState.OPEN_SEALED) :
-                                new BArrayType(resultType, arrayTypeSymbol, size, BArrayState.CLOSED_SEALED);
-            }
-            resultType = arrayTypeSymbol.type = arrType;
+                BLangExpression size = arrayTypeNode.sizes[i];
+                if (size.getKind() == NodeKind.LITERAL || size.getKind() == NodeKind.NUMERIC_LITERAL) {
+                    Integer sizeIndicator = (Integer) (((BLangLiteral) size).getValue());
+
+                    if (sizeIndicator == OPEN_ARRAY_INDICATOR) {
+                        arrType = new BArrayType(resultType,
+                                arrayTypeSymbol,
+                                sizeIndicator,
+                                BArrayState.UNSEALED);
+                    } else if (sizeIndicator == INFERRED_ARRAY_INDICATOR) {
+                        arrType = new BArrayType(resultType,
+                                arrayTypeSymbol,
+                                sizeIndicator,
+                                BArrayState.OPEN_SEALED);
+                    } else {
+                        arrType =  new BArrayType(resultType,
+                                arrayTypeSymbol,
+                                sizeIndicator,
+                                BArrayState.CLOSED_SEALED);
+                    }
```

Comment:
Since it is only the array state that changes, IMO we can improve readability here by doing something like

```suggestion
                    BArrayState arrayState;
                    if (sizeIndicator == OPEN_ARRAY_INDICATOR) {
                        arrayState = BArrayState.UNSEALED;
                    } else if (sizeIndicator == INFERRED_ARRAY_INDICATOR) {
                        arrayState = BArrayState.OPEN_SEALED;
                    } else {
                        arrayState = BArrayState.CLOSED_SEALED;
                    }
                    arrType =  new BArrayType(resultType, arrayTypeSymbol,  sizeIndicator, arrayState);
```